src/rgw/rgw_d3n_datacache.cc

```diff
@@ -0,0 +1,368 @@
+  ldout(cct, 20) << "D3nDataCache: Before eviction _free_data_cache_size:" << _free_data_cache_size << ", _outstanding_write_size:" << _outstanding_write_size << ", freed_size:" << freed_size << dendl;
+  while (len >= (_free_data_cache_size - _outstanding_write_size + freed_size)) {
+    ldout(cct, 20) << "D3nDataCache: enter eviction, r=" << r << dendl;
+    if (eviction_policy == _eviction_policy::LRU) {
+      r = lru_eviction();
+    } else if (eviction_policy == _eviction_policy::RANDOM) {
+      r = random_eviction();
+    } else {
+      ldout(cct, 0) << "D3nDataCache: Warning: unknown cache eviction policy, defaulting to lru eviction" << dendl;
+      r = lru_eviction();
+    }
+    if (r < 0)
+      return;
+    freed_size += r;
+  }
+  r = d3n_libaio_create_write_request(bl, len, oid);
+  if (r < 0) {
+    const std::lock_guard l(d3n_cache_lock);
+    auto it = d3n_outstanding_write_list.find(oid);
+    if (it != d3n_outstanding_write_list.end()) {
+      d3n_outstanding_write_list.erase(it);
+    }
+    ldout(cct, 1) << "D3nDataCache: create_aio_write_request fail, r=" << r << dendl;
+    return;
+  }
+
+  const std::lock_guard l(d3n_eviction_lock);
+  free_data_cache_size += freed_size;
+  outstanding_write_size += len;
+}
+
+bool D3nDataCache::get(const string& oid, const off_t len)
+{
+  const std::lock_guard l(d3n_cache_lock);
+  bool exist = false;
+  string location = cache_location + oid;
+
+  lsubdout(g_ceph_context, rgw_datacache, 20) << "D3nDataCache: " << __func__ << "(): location=" << location << dendl;
+  std::unordered_map<string, D3nChunkDataInfo*>::iterator iter = d3n_cache_map.find(oid);
+  if (!(iter == d3n_cache_map.end())) {
+    // check inside cache whether file exists or not!!!! then make exist true;
+    struct D3nChunkDataInfo* chdo = iter->second;
+    struct stat st;
+    int r = stat(location.c_str(), &st);
+    if ( r != -1 && st.st_size == len) { // file exists and containes required data range length
+      exist = true;
+      /*LRU*/
+      /*get D3nChunkDataInfo*/
+      const std::lock_guard l(d3n_eviction_lock);
+      lru_remove(chdo);
+      lru_insert_head(chdo);
+    } else {
+      d3n_cache_map.erase(oid);
+      const std::lock_guard l(d3n_eviction_lock);
+      lru_remove(chdo);
+      delete chdo;
+      exist = false;
+    }
+  }
+  return exist;
+}
+
+size_t D3nDataCache::random_eviction()
+{
+  lsubdout(g_ceph_context, rgw_datacache, 20) << "D3nDataCache: " << __func__ << "()" << dendl;
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  D3nChunkDataInfo* del_entry;
+  string del_oid, location;
+  {
+    const std::lock_guard l(d3n_cache_lock);
+    n_entries = d3n_cache_map.size();
+    if (n_entries <= 0) {
+      return -1;
+    }
+    srand (time(NULL));
+    random_index = ceph::util::generate_random_number<int>(0, n_entries-1);
+    std::unordered_map<string, D3nChunkDataInfo*>::iterator iter = d3n_cache_map.begin();
+    std::advance(iter, random_index);
```

Comment:
what are some alternatives?

Reply:
from previous discussion at https://github.com/ceph/ceph/pull/36266#discussion_r537906693:

> if we were to use unordered_map instead, maybe we could cheat and call cache_map.begin() random-ish?

another alternative would be to only support lru_eviction() and not random_eviction()

Reply:
apologizing for squashing confusion or my misunderstanding
it was changed to `std::unordered_map`

Reply:
`std::advance()` is still O(n) on `std::unordered_map` since it doesn't support random access

my suggestion was to cheat and just evict `std::unordered_map::begin()`, which is O(1). but i don't think that's a good solution; it isn't random, we'd just keep evicting entries with low hash values and never evict higher ones

a search for 'random entry in unordered_map' found [this on stackoverflow](https://stackoverflow.com/questions/27024269/select-random-element-in-an-unordered-map/27027557#27027557), which mentions the 'bucket interface' of unordered_map. so we could potentially select a random bucket in O(1), and get an iterator to the first entry in that bucket - but there's no guarantee that bucket has anything in it, so we may have to loop through buckets or keep sampling until we find something

@mkogan1 what do you think? by the time we need to evict, we're likely running full and O(n) operations will be really expensive. if we can't do better for random eviction, is it worth offering it as an option?